rp2040-hal/src/timer.rs

```diff
@@ -106,6 +112,61 @@ impl Timer {
 // safety: all write operations are synchronised and all reads are atomic
 unsafe impl Sync for Timer {}
 
+/// [`Timer`] based Delay
+pub struct Delay(());
+impl Delay {
+    fn delay_from_u32(&self, us: u32) {
+        // safety: only reads are made
+        let timer = unsafe { &*pac::TIMER::PTR };
+        let start = timer.timelr.read().bits();
+        while timer.timelr.read().bits().wrapping_sub(start) <= us {
+            // use nop instead of wfe because there is no event configured to wake us up.
+            cortex_m::asm::nop();
+        }
+    }
+
+    fn delay_from_u64(&self, us: u64) {
+        // safety: only reads are made
+        let timer = unsafe { &*pac::TIMER::PTR };
+        let start = get_counter(timer).ticks();
+        while get_counter(timer).ticks().wrapping_sub(start) <= us {
+            // use nop instead of wfe because there is no event configured to wake us up.
+            cortex_m::asm::nop();
+        }
+    }
+}
+impl embedded_hal::blocking::delay::DelayUs<u32> for Delay {
+    fn delay_us(&mut self, us: u32) {
+        self.delay_from_u32(us)
+    }
+}
+impl embedded_hal::blocking::delay::DelayUs<u64> for Delay {
+    fn delay_us(&mut self, us: u64) {
+        self.delay_from_u64(us)
+    }
+}
+impl embedded_hal::blocking::delay::DelayMs<u32> for Delay {
+    fn delay_ms(&mut self, ms: u32) {
+        for _ in 0..ms {
+            self.delay_from_u32(1000);
+        }
+    }
+}
+impl embedded_hal::blocking::delay::DelayMs<u64> for Delay {
```

Comment:
Implementing the trait for both u32 and u64 can cause a very strange error message:
```
error[E0277]: the trait bound `Delay: DelayMs<i32>` is not satisfied
  --> rp2040-hal/examples/blinky.rs:85:24
   |
85 |         delay.delay_ms(500);
   |               -------- ^^^ the trait `DelayMs<i32>` is not implemented for `Delay`
   |               |
   |               required by a bound introduced by this call
   |
   = help: the following other types implement trait `DelayMs<UXX>`:
             <Delay as DelayMs<u32>>
             <Delay as DelayMs<u64>>
```
This happens because rust doesn't know the exact type of the literal number `500`: All it knows it that it must be a number. Usually, that gets resolved automatically by type inference. But in this case, inference would be ambiguous: Either `u32` and `u64` could be valid. In that case, rust falls back to its default type: `i32`. Unfortunately, that's not a valid type for `DelayMs`, leading to the error message.

Of course, this can easily be fixed by calling `delay_ms(500u32)`. But that's a little bit ugly, and, more importantly, difficult to discover from the error message.

Reply:
Is there any recommendations on embedded-hal 0.2.7 w.r.t. this ? (I'll do an extra search for this)

Reply:
I don't know if there are recommendations, but I'm sure there are reasons e-h 1.0 went with only u32.
However, a quick search didn't bring up much besides "less code duplication" and "a blocking delay should never be >1h, so why care about u64?".

Reply:
This is what cortex-m does to avoid this paper cut: https://docs.rs/cortex-m/0.7.7/src/cortex_m/delay.rs.html#85

Reply:
Uh, and another issue: While playing with https://github.com/rust-embedded-community/embedded-sdmmc-rs/pull/80 I found this usage of `DelayUs<u8>`: https://github.com/rust-embedded-community/embedded-sdmmc-rs/pull/80/files#diff-cb7954c75c0cda9068329334e554ea5ce46a5ec8a40692a40be0e760f8c31e9fR37

If such things are common, we should implement the traits for all reasonable types, even if it causes a lot of duplication. Like cortex-m does.

Reply:
A code search on github actually found many instances of `DelayUs<u8>` and `DelayUs<u16>` that were not just implementations of those traits. Interestingly, nobody uses `DelayUs<u64>`.

Reply:
I added support for u8 to u64 and the i32 work around.
Do you think we should also cover for all the other signed integers ?

Reply:
No, signed types do not make sense here. Even i32 is very questionable, but it's necessary because rusts fallback to that type if it can't infer the exact type of an integer.